util.py

```diff
@@ -0,0 +1,359 @@
+        super().__init__(interval, function, args, kwargs)
+        self.stop = False
+
+    def cancel(self) -> None:
+        self.stop = True
+        super().cancel()
+
+    def reset(self, new_interval = None) -> None:
+        if new_interval is not None:
+            self.interval = new_interval
+        self.finished.set()
+
+    def run(self) -> None:
+        while not self.stop:
+            if not self.finished.wait(self.interval):
+                self.function(*self.args, **self.kwargs)
+            self.finished.clear()
+
+
+class Switch:
+    def __init__(
+            self,
+            key: Key,
+            pin: int,
+            pi_gpio: pigpio.pi,
+            callback: Callable[[Key, bool], None],
+            debounce: float = 0.05,
+            long_push_time: float = 0.5,
+            ) -> None:
+        """ Maintains current state of push button after debouncing.
+            Calls callback(key, False) when button is pushed.
+            Calls callback(key, True) again when button is pushed for more than long_push_time seconds
+        """
+        self.key = key
+        self.pin = pin
+        pi_gpio.set_mode(pin, pigpio.INPUT)
+        pi_gpio.set_pull_up_down(pin, pigpio.PUD_DOWN)
+        self.current_state = pi_gpio.read(pin)
+        self._lock = Lock()
+        self._debounce_timer = ResettableTimer(debounce, self.change_state)
+        self._debounce_timer.start()
+        self._long_timer = ResettableTimer(long_push_time, callback, self.key, True)
+        self._long_timer.start()
+        self._edge_callback = pi_gpio.callback(pin, pigpio.EITHER_EDGE, self.edge_change)
+        self.callback = callback
+
+    def edge_change(self, _1, level, _2) -> None:
+        with self._lock:
+            if level != 2:
+                if level != self.current_state:
+                    self._debounce_timer.reset()
+                else:
+                    self._debounce_timer.stop()
+
+    def change_state(self) -> None:
+        self.current_state = not self.current_state
+        if self.current_state:
+            self.callback(self.key, False)
+            self._long_timer.reset()
+        else:
+            self._long_timer.stop()
+
+    def clean(self) -> None:
+        """@brief Call when done using switch."""
+        self._edge_callback.cancel()
+        self._debounce_timer.cancel()
+        self._long_timer.cancel()
+        self._debounce_timer.join(1)
+        self._long_timer.join(1)
+
+
+class FileLock:
+    def __init__(self, lock_filepath: str) -> None:
+        self.depth = 0
+        self.lock_filepath = lock_filepath
+        self.lock_file: TextIOWrapper | None = None
+
+    def __enter__(self) -> None:
+        if self.depth == 0:
+            self.lock_file = open(self.lock_filepath, "w", encoding="utf-8")
+            try:
+                fcntl.flock(self.lock_file, fcntl.LOCK_EX)
+            except:
```

Comment:
`* as exc`?
Why do we catch exceptions here?
We won't have to close if we use with open.
Do we have to set self.lock_file to None if this program will be shutdown anyways?
Please consider.

Reply:
I don't use '* as exc' because I don't need to access this exception(it's reraised and logged later)
I catch exceptions just in case, flock can probably raise exception only when interrupted.
Using `with` - same answer as higher.
I probably don't need to set self.lock_file to None but I think it's better than holding closed file object.